### PR TITLE
Adding S3 to host static assets

### PR DESCRIPTION
**Pull request by jordanh**

### CODE REVIEW

**Comment by jordanh** [comment]
@mattkrick this is up on [staging](https://action-staging.parabol.co) at the moment. All of the JS files code-split by webpack work 👌 . 

The only thing I can't quite figure out is how to get references to our other static assets (i.e. png|jpg|jpeg|gif|svg|woff|woff2 or eot|ttf|wav|mp3) to be loaded from there too. Do I need to change the loader configuration?

File: .env.example [hunk 1]
```diff
@@ -14,6 +14,14 @@ AUTH0_CLIENT_SECRET="BksPeQQrRkXhDrugzQDg5Nw-IInub9RkQ-pSWohUM9s6Oii4xoGVCrK2_Oc
 AUTH0_DOMAIN="parabol.auth0.com"
 AUTH0_MANAGEMENT_TOKEN="eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJrRGZOTURJUXRxQ21lbVFGVHVoTjh6WVN4TlNFN3JjaSIsInNjb3BlcyI6eyJ1c2VycyI6eyJhY3Rpb25zIjpbInVwZGF0ZSJdfSwidXNlcnNfYXBwX21ldGFkYXRhIjp7ImFjdGlvbnMiOlsicmVhZCIsInVwZGF0ZSIsImRlbGV0ZSIsImNyZWF0ZSJdfX0sImlhdCI6MTQ3MTkxODAzNCwianRpIjoiMzdlOWFmYjRlMjEzOWYzMzNkMDkyNjY0NDBhMzk5MDEifQ.a6NANNe5IOzvFEomZm9OeeNpRfPjoMCUzA12PaK3zVg"
 
+# AWS S3
+AWS_ACCESS_KEY_ID="XXXXXXXXXXXXXXXXXXXX"
+AWS_REGION="us-west-2"
+AWS_S3_BUCKET="action-XXXXXXXX"
+AWS_SECRET_ACCESS_KEY="XXXXXXXXXXXXXXXXXXXXXXXX"
+
+CDN_URL="//s3-host.parabol.co"
```
Comment (jordanh) [inline .env.example]: we omit the protocol here, so that the client amends it for us.

File: src/client/webpackEntry.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Setup webpack runtime public path loading, used for configuring
+ * asset loading from our CDN:
+ */
+
+/* eslint-disable camelcase, no-undef */
+const getWebpackPublicPath = require('universal/utils/getWebpackPublicPath');
+__webpack_public_path__ = getWebpackPublicPath();
```
Comment (jordanh) [inline src/client/webpackEntry.js]: N.B. we must put them in their own entry file. When I added this statement to `client.js` the `import` statements were getting hosted above the `__webpack_public_path__ = ` expression, and then the magic ✨  didn't work 🚫 .

File: webpack/webpack.config.prod.js [hunk 3]
```diff
@@ -31,22 +32,31 @@ const deployPlugins = [];
 if (process.env.DEPLOY) {
   deployPlugins.push(new webpack.optimize.UglifyJsPlugin({compressor: {warnings: false}, comments: /(?:)/}));
   deployPlugins.push(new webpack.LoaderOptionsPlugin({comments: false}));
+  deployPlugins.push(new S3Plugin({
```
Comment (jordanh) [inline webpack/webpack.config.prod.js]: This plugin is pretty great. It'll upload all of our static assets to S3 for us as part of our webpack process.

File: webpack/webpack.config.prod.js [hunk 4]
```diff
@@ -62,7 +72,11 @@ export default {
     }),
     new webpack.optimize.AggressiveMergingPlugin(),
     new webpack.optimize.MinChunkSizePlugin({minChunkSize: 50000}),
-    new AssetsPlugin({path: path.join(root, 'build'), filename: 'assets.json', includeManifest: true}),
+    new AssetsPlugin({
```
Comment (jordanh) [inline webpack/webpack.config.prod.js]: N.B. only change here was to reflow this.

**Comment by mattkrick** [comment]
oh wowwwwwww! so we run webpack, it sends everything to S3, sets S3 as our /static/,and then our node server doesn't have do serve up a single doc? so we could nuke the node static server in another PR? It looks good to me. This woulda taken me for ever to figure out!

**Comment by jordanh** [comment]
@mattkrick that's right – it posts everything to S3 and then eliminates node from having to serve anything from `/static`. I suggest we leave the `/static` route in node for a bit. If anything gets hosed we should be able to unset the `CDN_URL` environment variable and our application will go back to its previous behavior... (I am going to test this now on staging after I merge)

**Pull request merged**